### PR TITLE
Make database init / update transactional

### PR DESCRIPTION
**Pull request by kim**
# Description of Changes

That is, run both schema modifications as well as init/update reducers in the same transaction.

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

### CODE REVIEW

**Comment by kim** [comment]
It's not _ideal_ having to let go of ownership of the `MutTxId` when calling the reducers -- I'd rather like to get it back in case something else needs to run in the same tx.

File: crates/core/src/host/wasm_common/module_host_actor.rs [hunk 1]
```diff
@@ -504,8 +452,59 @@ impl<T: WasmInstance> ModuleInstance for WasmModuleInstance<T> {
 }
 
 impl<T: WasmInstance> WasmModuleInstance<T> {
+    fn call_reducer_internal(
```
Comment (gefjon) [inline crates/core/src/host/wasm_common/module_host_actor.rs:455]: Could you add a doc comment describing the responsibilities of this method? In particular, I think it's non-obvious that it commits the transaction on success or cancels it on failure.

File: crates/core/src/host/wasm_common/module_host_actor.rs [hunk 2]
```diff
@@ -708,6 +707,63 @@ impl<T: WasmInstance> WasmModuleInstance<T> {
...
+                    index.table_id = known_schema.table_id;
+                }
+                let known_schema = TableDef::from(known_schema);
+                if known_schema != proposed_schema {
+                    self.system_logger()
+                        .warn(&format!("stored and proposed schema of `{}` differ", table.name));
+                    tainted_tables.push(table.name.to_owned());
+                } else {
+                    // Table unchanged
+                }
```
Comment (gefjon) [inline crates/core/src/host/wasm_common/module_host_actor.rs:740]:
```suggestion
                // If the schema has changed, the table is tainted.
                // If the schemas match, everything is peachy and we can continue.
                if known_schema != proposed_schema {
                    self.system_logger()
                        .warn(&format!("stored and proposed schema of `{}` differ", table.name));
                    tainted_tables.push(table.name.to_owned());
                }
```

Empty `else` block is a code smell to me. I'd rather see a comment introducing the conditional itself.

**Review by gefjon — COMMENTED** [review]
I'm not terribly familiar with this code, but from what I've read this new version seems more clear; I'm a fan. (It also seems more semantically correct, but I don't think that's really in question.)

I've left a few minor requests for comments, but largely I found this code pleasantly self-describing.

**Review by gefjon — APPROVED** [review]

File: crates/core/src/host/wasm_common/module_host_actor.rs (hunk 2 above)
Comment (kim) [reply crates/core/src/host/wasm_common/module_host_actor.rs]: c89b358816c116a729f4a0c1096c3ed581cb68ca

File: crates/core/src/host/wasm_common/module_host_actor.rs (hunk 1 above)
Comment (kim) [reply crates/core/src/host/wasm_common/module_host_actor.rs:470]: c89b358816c116a729f4a0c1096c3ed581cb68ca

I didn't author the code, only moved it around, so this basically reflects my understanding from just reading the code. Also left a TODO regarding why it is that we can't get back our transaction easily.

**Pull request merged**